### DISCUSSION TITLE
How can I get a source of a symbol?

### DISCUSSION THREAD

**Discussion by saruman9** [Vector35/binaryninja-api#3020] [API] [answered]
For example, I have user defined functions, exported functions, functions with auto names (via analysis), etc. How can I get only functions with auto names for further renaming? So this question is the same for all: symbols, types, variables, etc. See [Ghidra's API](https://ghidra.re/ghidra_docs/api/ghidra/program/model/symbol/SourceType.html#enum.constant.summary) for more explanation for what I does mean. I know about `auto` property, but I'm not sure if that's what I need and that's enough for my cases.

**Comment by psifertex** [comment, marked as answer]
So you only want things like sub_ that don't have a debug symbol or a user symbol?

**Reply by saruman9** [reply]
I want all functions, which have auto names: `sub_*`, demangled, functions from import/export, etc. I want to rename and set signature/prototype for functions, which have not been modified by me.

**Reply by psifertex** [reply]
So you want to check whether the _symbol_ at that function is an auto symbol. Only the functions you have manually renamed will have user symbols:

```
for f in bv.functions:
  if f.symbol.auto:
    continue
  # Do whatever here for functions the user has renamed
  print(f)
```

**Reply by saruman9** [reply]
Oh, `f.symbol.auto` not `f.auto`! Thank you!

**Comment by psifertex** [comment]
Another related question that folks might be asking and find this question is:

`How do I differentiate between sub_ symbols and "real" symbols made by debug information? `
They both are auto symbols (as shown above) but one is real and the other isn't.

The short explanation is that if you access a function.symbol and a symbol doesn't exist, a temporary one with `sub_OFFSET` is created on the fly. That said, the symbol doesn't really exist and so if you instead use:

```
bv.get_symbol_at(function.start)
``` 
it will return None for sub_ named functions (without having to string match or in case some debug symbols happen to start with that)

**Reply by saruman9** [reply]
It's also useful for my cases. Thank you!
